modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java

```diff
@@ -346,6 +347,11 @@ static Query toQuery(QueryShardContext context, boolean mapUnmappedFieldsAsStrin
         // as an analyzed string.
         context.setAllowUnmappedFields(false);
         context.setMapUnmappedFieldAsString(mapUnmappedFieldsAsString);
+        // From 5.0 and onwards the percolator rewrites the query at index time,
+        // this is not the case for percolator queries in indices created before 5.0
+        if (context.getIndexSettings().getIndexVersionCreated().before(Version.V_5_0_0_alpha1)) {
+            queryBuilder = queryBuilder.rewrite(context);
```

Comment:
Why not QueryBuilder#rewriteQuery ? Otherwise you can still have non-primitive queries ? 
Also, we should rewrite query rewritten at index time (post v5) too ? Otherwise all rewritten rules must be the same at index and query time ?